diff --git a/meson.build b/meson.build
index dbed6ee..c9d0fd8 100644
--- a/meson.build
+++ b/meson.build
@@ -671,8 +671,8 @@ if with_glx != 'disabled'
   elif with_glx == 'xlib'
     if not with_gallium
       error('xlib based GLX requires at least one gallium driver')
-    elif not with_gallium_swrast
-      error('xlib based GLX requires softpipe or llvmpipe.')
+    elif not with_gallium_swrast and not with_gallium_virgl
+      error('xlib based GLX requires softpipe, llvmpipe, or virgl.')
     elif with_dri
       error('xlib conflicts with any dri driver')
     endif
diff --git a/src/gallium/auxiliary/hud/hud_context.c b/src/gallium/auxiliary/hud/hud_context.c
index cc43d3b..02eff74 100644
--- a/src/gallium/auxiliary/hud/hud_context.c
+++ b/src/gallium/auxiliary/hud/hud_context.c
@@ -1799,7 +1799,8 @@ hud_set_draw_context(struct hud_context *hud, struct cso_context *cso,
    {
       static const char *vertex_shader_text = {
          "VERT\n"
-         "DCL IN[0..1]\n"
+         "DCL IN[0]\n"
+         "DCL IN[1]\n"
          "DCL OUT[0], POSITION\n"
          "DCL OUT[1], COLOR[0]\n" /* color */
          "DCL OUT[2], GENERIC[0]\n" /* texcoord */
@@ -1842,7 +1843,8 @@ hud_set_draw_context(struct hud_context *hud, struct cso_context *cso,
        * to match the varyings in fs_text */
       static const char *vertex_shader_text = {
          "VERT\n"
-         "DCL IN[0..1]\n"
+         "DCL IN[0]\n"
+         "DCL IN[1]\n"
          "DCL OUT[0], POSITION\n"
          "DCL OUT[1], GENERIC[0]\n" /* texcoord */
          /* [0] = color,
diff --git a/src/gallium/auxiliary/target-helpers/inline_sw_helper.h b/src/gallium/auxiliary/target-helpers/inline_sw_helper.h
index ddb67b3..2cd0257 100644
--- a/src/gallium/auxiliary/target-helpers/inline_sw_helper.h
+++ b/src/gallium/auxiliary/target-helpers/inline_sw_helper.h
@@ -39,7 +39,7 @@ sw_screen_create_named(struct sw_winsys *winsys, const char *driver)
       screen = llvmpipe_create_screen(winsys);
 #endif
 
-#if defined(GALLIUM_VIRGL)
+#ifdef GALLIUM_VIRGL
    if (screen == NULL && strcmp(driver, "virpipe") == 0) {
       struct virgl_winsys *vws;
       vws = virgl_vtest_winsys_wrap(winsys);
@@ -75,6 +75,9 @@ sw_screen_create_vk(struct sw_winsys *winsys, bool sw_vk)
 #if defined(GALLIUM_D3D12)
       (sw_vk || only_sw) ? "" : "d3d12",
 #endif
+#if defined(GALLIUM_VIRGL)
+      (sw_vk ? "" : "virpipe"),
+#endif
 #if defined(GALLIUM_LLVMPIPE)
       "llvmpipe",
 #endif
diff --git a/src/gallium/auxiliary/vl/vl_winsys_xlib_swrast.c b/src/gallium/auxiliary/vl/vl_winsys_xlib_swrast.c
index 1d3b4a8..2674583 100644
--- a/src/gallium/auxiliary/vl/vl_winsys_xlib_swrast.c
+++ b/src/gallium/auxiliary/vl/vl_winsys_xlib_swrast.c
@@ -150,7 +150,8 @@ vl_swrast_texture_from_drawable(struct vl_screen *vscreen, void *drawable)
 
    XWindowAttributes x11_window_attrs = {};
    XGetWindowAttributes(scrn->display, x11_window, &x11_window_attrs);
-   enum pipe_format x11_window_format = vl_dri2_format_for_depth(&scrn->base, x11_window_attrs.depth);
+   enum pipe_format x11_window_format = x11_window_attrs.depth == 24 ?
+      PIPE_FORMAT_B8G8R8X8_UNORM : PIPE_FORMAT_B8G8R8A8_UNORM;
 
    bool needs_new_back_buffer_allocation = true;
    if (scrn->drawable_texture) {
